src/components/Button/withOnSubmitDisabling.js

```diff
@@ -0,0 +1,36 @@
+// @flow
+import * as React from 'react'
+import { h, Component } from 'preact'
+
+type State = {
+  disabled: boolean,
+}
+
+type Props = {
+  disabled: ?boolean,
+  shouldBeDisabledOnClick: ?boolean,
+  onClick: Function,
+}
+
+export default (WrappedButton: React.ComponentType<Props>): React.ComponentType<Props> =>
+  class WithOnSubmitDisabling extends Component<Props, State> {
+
+  constructor() {
+    super()
+    this.state = {
+      disabled: false
+    }
+  }
+
+  handleClick = () => {
+    if (this.props.shouldBeDisabledOnClick) {
+      this.setState({ disabled: true })
```

Comment:
I don't get one thing - seems like you disable the button, but do not enable it back after successful call?

Reply:
Yes. These buttons trigger a screen change. So once you click on them you should be taken to a new screen and the button becomes disabled. If you visit the screen again it will be enabled

Reply:
I think with such approach it's easy to make a bug at some later stage.

One of the option would be to have `onClick` prop to be a promise, resolving/rejecting it would re-enable the button back. This IMO would be ideal.

Another case (less clean but still OK) would be to call `onClick` with a callback that would be executed outside of this HOC e.g. when you know the call has been completed.

In any case, I did approve this PR only because it satisfies current requirements.

Reply:
I feel letting the parent component's state determine if the button should be disabled is a better approach. That way we keep the button component as a simple controlled component and stateless.

Reply:
This is also introducing 2 different ways the Button component could be disabled. Either with a prop or its own state.

Reply:
Actually, I like @josokinas suggestion to make use of a promise now that I've caught up with the conversation.

Reply:
Yes I sort of agree @kopijunkie, usually having something being controlled both with external prop and internally is not a good idea...